File: project_euler/problem019.py
# You are given the following information, but you may prefer to do some research for yourself.
#
# 1 Jan 1900 was a Monday.
# Thirty days has September,
# April, June and November.
# All the rest have thirty-one,
# Saving February alone,
# Which has twenty-eight, rain or shine.
# And on leap years, twenty-nine.
# A leap year occurs on any year evenly divisible by 4, but not on a century unless it is divisible by 400.
# How many Sundays fell on the first of the month during the twentieth century (1 Jan 1901 to 31 Dec 2000)?

def problem019():
    n_first_month_sundays = 0  # We'll increment this to keep track
    day_of_week = 0  # We'll increment this mod 7 every day. 0 == monday, ..., 6 == sunday
    for year in range(1900, 2001):

        # Determine if this is a leap year
        is_leap_year = False
        if year % 100 == 0:  # if a century, its only a leap year if its divisible by 400
            is_leap_year = (year % 400 == 0)
        else:
            is_leap_year = (year % 4 == 0)

        for month in range(1, 13):

            # Figure out how many days are in this month
            n_days_in_month = 31
            if month in [4, 6, 9, 11]:
                n_days_in_month = 30
            if month == 2:
                if is_leap_year:
                    n_days_in_month = 29
                else:
                    n_days_in_month = 28

            for day_of_month in range(1, n_days_in_month + 1):

                if (year >= 1901) and (year <= 2000) and (day_of_month == 1) and (day_of_week == 6):
                    n_first_month_sundays += 1

                # increment the day of the week mod 7
                day_of_week = (day_of_week + 1) % 7

    return n_first_month_sundays

File: R/R/problem019.R
# Problem 19: Counting Sundays
# You are given the following information, but you may prefer to do some research for yourself.
# - 1 Jan 1900 was a Monday.
# - Thirty days has September,
#   April, June and November.
#   All the rest have thirty-one,
#   Saving February alone,
#   Which has twenty-eight, rain or shine.
#   And on leap years, twenty-nine.
# - A leap year occurs on any year evenly divisible by 4, but not on a century unless it is divisible by 400.
# How many Sundays fell on the first of the month during the twentieth century (1 Jan 1901 to 31 Dec 2000)?

problem019 <- function(){

  # Jan 1 1900 was a monday.
  # 1900 wasn't a leap year (not divisible by 400), so Jan 1 1901 was 365 days
  # later, making it a tuesday (365 % 7 == 1, so the day of week advanced by 1).

  days <- 0 # Day 0 is Jan 1 1901, a Tuesday.

  sundays_on_the_1st <- 0

  # Start counting month by month. From here out out, any day such that day %% 7 == 5 is a sunday.
  for(year in 1901:2000){
    is_leap_year <- (year %% 4 == 0 & year %% 100 != 0) | (year %% 400 == 0)

    for(month in 1:12){

      if(days %% 7 == 5){
        sundays_on_the_1st <- sundays_on_the_1st + 1
      }

      if(month %in% c(9,4,6,11)){
        days <- days + 30
      } else if(month == 2) {
        days <- days + ifelse(is_leap_year,29,28)
      } else {
        days <- days + 31
      }
    }
  }
  sundays_on_the_1st
}
